src/transport_udp.cpp

```diff
@@ -242,361 +520,480 @@ UDPTransport::fd_writer()
  *  - Lookup addr in map to find context and name info
  *  - If context doesn't exist, then it's a new connection and the delegate will
  * be called after creating new context
- *  - Create connData and send to queue
+ *  - Create ConnData and send to queue
  *  - Call on_recv_notify() delegate to notify of new data available. This is
  * not called again if there is still pending data to be dequeued for the same
  * StreamId
  */
-void
-UDPTransport::fd_reader()
-{
-  logger->Log("Starting transport reader thread");
-
-  const int dataSize = 65535; // TODO Add config var to set this value.  Sizes
-                              // larger than actual MTU require IP frags
-  struct sockaddr_storage remoteAddr;
-  memset(&remoteAddr, 0, sizeof(remoteAddr));
-  socklen_t remoteAddrLen = sizeof(remoteAddr);
-
-  uint8_t data[dataSize];
-
-  while (not stop) {
-    int rLen = recvfrom(fd,
-                        data,
-                        dataSize,
-                        0 /*flags*/,
-                        (struct sockaddr*)&remoteAddr,
-                        &remoteAddrLen);
-
-    if (rLen < 0 || stop) {
-      if ((errno == EAGAIN) || (stop)) {
-        // timeout on read or stop issued
-        continue;
-
-      } else {
-        logger->error << "Error reading from UDP socket: " << strerror(errno)
-                      << std::flush;
-        break;
-      }
+void UDPTransport::fd_reader() {
+    logger->Log("Starting transport reader thread");
+
+    const int dataSize = 65535; // TODO Add config var to set this value.  Sizes
+    // larger than actual MTU require IP frags
+    uint8_t data[dataSize];
+
+    while (not stop) {
+        Addr remote_addr;
+
+        int rLen = recvfrom(fd,
+                            data,
+                            dataSize,
+                            0 /*flags*/,
+                            (struct sockaddr *) &remote_addr.addr,
+                            &remote_addr.addr_len);
+
+        if (rLen < 0 || stop) {
+            if ((errno == EAGAIN) || (stop)) {
+                // timeout on read or stop issued
+                continue;
+
+            } else {
+                logger->error << "Error reading from UDP socket: " << strerror(errno)
+                              << std::flush;
+                break;
+            }
+        }
+
+        if (rLen == 0) {
+            continue;
+        }
+
+        if (data[0] != UdpProtocol::PROTOCOL_VERSION) {
+            // TODO: Add metrics to track discards on invalid received message
+            continue;
+        }
+
+        const auto current_tick = _tick_service->get_ticks(std::chrono::milliseconds(1));
+
+        remote_addr.id = create_addr_id(remote_addr.addr);
+        const auto a_conn_it = addr_conn_contexts.find(remote_addr.id);
```

Comment:
Isn't a lock needed to protect access to `conn_contexts`?

Reply:
Yes, thanks. I've added that.